README.md
# Json Parse [![Build Status](https://travis-ci.org/mitchhentges/json-parse.svg?branch=master)](https://travis-ci.org/mitchhentges/json-parse)

A tool to quickly parse JSON into Java maps and lists. Significantly faster than FasterXML's
Jackson with an extremely small memory and size footprint.

_Lower is better:_

![](resources/comparison-v1.3.4.png)

[Wiki page explaining source of results](https://github.com/mitchhentges/json-parse/wiki/Comparison-with-FasterXML-Jackson)

## Usage

```
String mapString = "{\"fast\":true, \"super-neat\":true}";
String listString = "[1, 2, false]";
String stringString = "\"yo dawg\"";
String numberString = "-1.2e-3";
String boolString = "true";

Map<String, Object> map = JsonParse.map(mapString);
List<Object> list = JsonParse.list(listString);
String string = JsonParse.string(stringString);
Number number = JsonParse.number(numberString);
Boolean bool = JsonParse.bool(boolString);

// Will still parse if type isn't explicitly specified
Object unknown = JsonParse.parse(scanner.nextLine());
System.out.println(unkown.getClass());
```

## Getting the dependency

**Maven**
```
<dependency>
    <groupId>ca.fuzzlesoft</groupId>
    <artifactId>json-parse</artifactId>
    <version>1.3.4</version>
</dependency>
```

**Gradle**
```
compile 'ca.fuzzlesoft:json-parse:1.3.4'
```

## Features

### Shows path to syntax error if given invalid JSON

If incorrect JSON is passed to `JsonParse`, the thrown exception will explain _where_ the JSON is invalid.

```
{"outer": {"inner": {"a": true, "b": nulll}}}
<root>.outer.inner.b: "nulll" is not a valid constant. Missing quotes?

{"outer": {"inner": {"a": true "b": false}}}
<root>.outer.inner.a: wasn't followed by a comma

{"outer": {"no-colon" true}}
<root>.outer.no-colon: "no-colon" wasn't followed by a colon

{"object": {no-quotes: true}}
<root>.object: unexpected character 'n' where a property name is expected. Missing quotes?

{"outer": {"sick-list": [{}, {"inner": [1, [1, 2, 3, [fasle]]]}]}}
<root>.outer.sick-list.[1].inner.[1].[3].[0]: "fasle" is not a valid constant. Missing quotes?
```


### Thread safe

Can be used from multiple thread contexts without losing data integrity or failing.

```
public static void main(String[] args) throws InterruptedException {
    List<Thread> threads = new ArrayList<>();
    for (int i = 0; i < 10; i++) {
        final int id = i;
        Thread thread = new Thread(new Runnable() {
            @Override
            public void run() {
                try {
                    List<Object> ints = JsonParse.list("[1, 2, 3]");
                    System.out.printf("%d: %d,%d,%d\n", id, ints.get(0), ints.get(1), ints.get(2));
                } catch (JsonParseException e) {
                    System.err.println("Exception thrown");
                }
            }
        });
        thread.start();
        threads.add(thread);
    }
    for (int i = 0; i < 10; i++) {
        threads.get(i).join();
    }
}
```

```
0: 1,2,3
9: 1,2,3
8: 1,2,3
7: 1,2,3
5: 1,2,3
6: 1,2,3
1: 1,2,3
3: 1,2,3
2: 1,2,3
4: 1,2,3
```

### Convert JSON strings to a Java objects.


```
// For example, arrays.
String listString = "[1, 2, false]";
List<Object> list = JsonParse.list(listString);
```

## FAQ

* Can this convert from JSON directly to Plain Old Java Objects?

Unfortunately not. JSON doesn't have any type information, so explicit knowledge is required before JSON can be
converted. That would be the job of a "binding framework"

* Why is the code squashed so much?

I wanted to minimize the number of function calls. Though the cost is minimal, it will still affect performance
on the large scale. Besides, my goal is for this to be a "finish and forget" project, where the speed benefits
add up over time, while the maintenance cost is nonexistent because the code "just works". We'll see how naive this
is over time ;)

## License
[MIT License (Expat)](http://www.opensource.org/licenses/mit-license.php)


CHANGELOG.md
# 1.3.4

Bugfix release

* Fix formatting of some JsonParseException errors, such as "Unexpected character ... instead of array value"
* Properly handles escape sequences and unicode code points
* Fixes issue where if a string ended with an escaped double-quote, a failure would be encountered

# 1.3.3

Bugfix release

* Change format of error-trace message
* Failing to parse root constant should throw JsonParseException, not NullPointerException
* Ignore all input after root element

# 1.3.2

Bugfix release

* Empty strings will no longer break parsing

# 1.3.1

Performance release

* Now will handle arrays faster, as it delimits their current "propertyName" as `null` instead of `"[]"`

# 1.3.0

Performance and array-index-trace release

* Array index is now shown in JSON traces. [0] is the first element, [1] is the second, and so on
* Can now parse strings, numbers and constants as the root element. It doesn't have to be an array or object anymore
* More internal consistency in the way root objects are handled
* Performance increased by at least 25% (only one stack is used for managing state)

# 1.2.0

* `list()` and `map()` are now static methods. Building an object is no longer necessary

# 1.1.0

Performance release

* Refactored object `NAME` and value types `STRING`, `NUMBER` and `CONSTANT` to use forward look-aheads (my term) to
increase speed

# 1.0.3

Performance release

* Detecting constants (whitespace, numbers) is now faster
* Improves avoidance of processing for end-of-block (e.g. object or array), when no end-of-block is occurring

# 1.0.2

* Don't include Junit in compile scope. Instead, it should only be available for testing

# 1.0.1

Performance release

* Don't check if non-string value is null/true/false up front, just attempt to parse it upon value termination
* Add `CHANGELOG.md` to see version changes

# 1.0.0

Initial release

* Convert json string to array/map.
* Show json stack trace on error

src/test/java/ca/fuzzlesoft/ContainingStripTest.java
package ca.fuzzlesoft;

import org.junit.Assert;
import org.junit.Test;

/**
 * @author mitch
 * @since 30/12/15
 */
public class ContainingStripTest {
    @Test
    public void shouldStripContainingObjectAndWhitespace() {
        String test = "     {\"a\":\"true\",\"b\":[{},{}]}  ";
        String expected = "\"a\":\"true\",\"b\":[{},{}]";
        Assert.assertEquals(expected, ContainingStrip.OBJECT.strip(test));
    }

    @Test
    public void shouldStripContainingArray() {
        String test = "[1,2]";
        String expected = "1,2";
        Assert.assertEquals(expected, ContainingStrip.ARRAY.strip(test));
    }

    @Test
    public void shouldThrowExceptionIfNoObject() {
        String noObject = "    ";
        String noEnd = " { ";
        String noStart = " } ";
        ContainingStrip strip = ContainingStrip.OBJECT;

        try {
            strip.strip(noObject);
            Assert.fail("Did not not throw exception when there was no containing object");
        } catch (Exception ignored) {}

        try {
            strip.strip(noEnd);
            Assert.fail("Did not throw exception when object didn't end");
        } catch (Exception ignored) {}

        try {
            strip.strip(noStart);
            Assert.fail("Did not throw exception when object didn't start");
        } catch (Exception ignored) {}
    }

    @Test
    public void shouldAllowTabsOrNewlinesAroundObject() {
        String test = "\t\n{\t}\n\t";
        String expected = "\t";
        Assert.assertEquals(expected, ContainingStrip.OBJECT.strip(test));
    }
}

src/main/java/ca/fuzzlesoft/ContainingStrip.java
package ca.fuzzlesoft;

/**
 * @author mitch
 * @since 30/12/15
 */
public class ContainingStrip {
    public static final ContainingStrip OBJECT = new ContainingStrip('{', '}');
    public static final ContainingStrip ARRAY = new ContainingStrip('[', ']');

    private final char toStripStart;
    private final char toStripEnd;

    public ContainingStrip(char toStripStart, char toStripEnd) {
        this.toStripStart = toStripStart;
        this.toStripEnd = toStripEnd;
    }

    public String strip(String string) {
        int beginning, i = 0;
        char current;

        //reach and remove first {
        while ((current = string.charAt(i++)) != toStripStart) {
            if (i >= string.length() || !Constants.is(Constants.WHITE, current)) {
                throw new JsonParseException("JSON string did not start with " + toStripStart);
            }
        }
        beginning = i;
        i = string.length() - 1;

        // reach and remove last }
        while ((current = string.charAt(i--)) != toStripEnd) {
            if (i < 0 || !Constants.is(Constants.WHITE, current)) {
                throw new JsonParseException("JSON string did not end with " + toStripEnd);
            }
        }
        return string.substring(beginning, i + 1);
    }
}


src/main/java/ca/fuzzlesoft/JsonParseException.java
package ca.fuzzlesoft;

import ca.fuzzlesoft.JsonParse.State;

import java.util.List;
import java.util.Stack;

/**
 * @author mitch
 * @since 30/12/15
 */
public class JsonParseException extends RuntimeException {

    private final String message;

    public JsonParseException(String message) {
        this.message = message;
    }

    public JsonParseException(Stack<State> stateStack, String message) {
        String jsonTrace = "";
        for (int i = 0; i < stateStack.size(); i++) {
            String name = stateStack.get(i).propertyName;
            if (name == null) {
                // Fill in array index
                List<Object> list = (List<Object>) stateStack.get(i).container;
                name = String.format("[%d]", list.size());
            }
            jsonTrace += name + (i != stateStack.size() - 1 ? "." : "");
        }

        jsonTrace = jsonTrace.equals("") ? "<root>" : "<root>." + jsonTrace;

        this.message = jsonTrace + ": " + message;
    }

    @Override
    public String getMessage() {
        return message;
    }
}


src/main/java/ca/fuzzlesoft/JsonParse.java
package ca.fuzzlesoft;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.IOException;
import java.util.*;

/**
 * Parses JSON, converting it into {@link List}s and {@link Map}s. Is thread safe.
 *
 * @author mitch
 * @since 30/12/15
 */
@SuppressWarnings("unchecked") // Because of reusing `currentContainer` for both maps and lists
public class JsonParse {

    private JsonParse() {}

    /**
     * Converts jsonString into a {@link Map}
     * @param jsonString parsed
     * @return the contents of the jsonString
     */
    public static Map<String, Object> map(String jsonString) {
        return (Map<String, Object>) parse(jsonString);
    }

    /**
     * Converts jsonString into a {@link List}
     * @param jsonString parsed
     * @return the contents of the jsonString
     */
    public static List<Object> list(String jsonString) {
        return (List<Object>) parse(jsonString);
    }

    /**
     * Pulls the internal JSON string from jsonString and returns it
     * @param jsonString parsed
     * @return the contents of the jsonString
     */
    public static String string(String jsonString) {
        return (String) parse(jsonString);
    }

    /**
     * Converts jsonString into a {@link Number}, be it an integer or floating-point
     * @param jsonString parsed
     * @return the contents of the jsonString
     */
    public static Number number(String jsonString) {
        return (Number) parse(jsonString);
    }

    /**
     * Converts jsonString into a boolean
     * @param jsonString parsed
     * @return the contents of the jsonString
     */
    public static boolean bool(String jsonString) {
        return (boolean) parse(jsonString);
    }

    /**
     * Parses jsonString according to what the outermost structure is
     * @param jsonString parsed
     * @return the contents of jsonString
     */
    @SuppressWarnings("ConstantConditions")
    public static Object parse(String jsonString) {
        Stack<State> stateStack = new Stack<>();
        Type currentType;

        boolean expectingComma = false, expectingColon = false;
        int fieldStart = 0, end = jsonString.length() - 1, i = 0;
        String propertyName = null;
        Object currentContainer = null;
        Object value;
        char current;

        try {
            while (Constants.isWhitespace((current = jsonString.charAt(i)))) i++;
        } catch (IndexOutOfBoundsException e) {
            throw new JsonParseException("Provided JSON string did not contain a value");
        }

        if (current == '{') {
            currentType = Type.OBJECT;
            currentContainer = new HashMap<>();
            i++;
        } else if (current == '[') {
            currentType = Type.ARRAY;
            currentContainer = new ArrayList<>();
            propertyName = null;
            i++;
        } else if (current == '"') {
            currentType = Type.STRING;
            fieldStart = i;
        } else if (Constants.isLetter(current)) {
            // Assume parsing a constant ("null", "true", "false", etc)
            currentType = Type.CONSTANT;
            fieldStart = i;
        } else if (Constants.isNumberStart(current)) {
            currentType = Type.NUMBER;
            fieldStart = i;
        } else {
            throw new JsonParseException(stateStack, "Unexpected character \"" + current + "\" instead of root value");
        }

        while (i <= end) {
            current = jsonString.charAt(i);
            switch (currentType) {
                case NAME:
                    try {
                        ExtractedString extracted = extractString(jsonString, i);
                        i = extracted.sourceEnd;
                        propertyName = extracted.str;
                    } catch (StringIndexOutOfBoundsException e) {
                        throw new JsonParseException(stateStack, "String did not have ending quote");
                    }
                    currentType = Type.HEURISTIC;
                    expectingColon = true;
                    i++;
                    break;
                case STRING:
                    try {
                        ExtractedString extracted = extractString(jsonString, i);
                        i = extracted.sourceEnd;
                        value = extracted.str;
                    } catch (StringIndexOutOfBoundsException e) {
                        throw new JsonParseException(stateStack, "String did not have ending quote");
                    }

                    if (currentContainer == null) {
                        return value;
                    } else {
                        expectingComma = true;
                        if (currentContainer instanceof Map) {
                            ((Map<String, Object>) currentContainer).put(propertyName, value);
                            currentType = Type.OBJECT;
                        } else {
                            ((List<Object>) currentContainer).add(value);
                            currentType = Type.ARRAY;

                        }
                    }

                    i++;
                    break;
                case NUMBER: {
                    boolean withDecimal = false;
                    boolean withE = false;
                    do {
                        current = jsonString.charAt(i);
                        if (!withDecimal && current == '.') {
                            withDecimal = true;
                        } else if (!withE && (current == 'e' || current == 'E')) {
                            withE = true;
                        } else if (!Constants.isNumberStart(current) && current != '+') {
                            break;
                        }
                    } while (i++ < end);

                    String valueString = jsonString.substring(fieldStart, i);
                    try {
                        if (withDecimal || withE) {
                            value = Double.valueOf(valueString);
                        } else {
                            value = Long.valueOf(valueString);
                        }
                    } catch (NumberFormatException e) {
                        throw new JsonParseException(stateStack, "\"" + valueString +
                                "\" expected to be a number, but wasn't");
                    }

                    if (currentContainer == null) {
                        return value;
                    } else {
                        expectingComma = true;
                        if (currentContainer instanceof Map) {
                            ((Map<String, Object>) currentContainer).put(propertyName, value);
                            currentType = Type.OBJECT;
                        } else {
                            ((List<Object>) currentContainer).add(value);
                            currentType = Type.ARRAY;
                        }
                    }
                    break;
                }
                case CONSTANT:
                    while (Constants.isLetter(current) && i++ < end) {
                        current = jsonString.charAt(i);
                    }

                    String valueString = jsonString.substring(fieldStart, i);
                    switch (valueString) {
                        case "false":
                            value = false;
                            break;
                        case "true":
                            value = true;
                            break;
                        case "null":
                            value = null;
                            break;
                        default:
                            if (currentContainer instanceof Map) {
                                stateStack.push(new State(propertyName, currentContainer, Type.OBJECT));
                            } else if (currentContainer instanceof List) {
                                stateStack.push(new State(propertyName, currentContainer, Type.ARRAY));
                            }
                            throw new JsonParseException(stateStack, "\"" + valueString
                                    + "\" is not a valid constant. Missing quotes?");
                    }

                    if (currentContainer == null) {
                        return value;
                    } else {
                        expectingComma = true;
                        if (currentContainer instanceof Map) {
                            ((Map<String, Object>) currentContainer).put(propertyName, value);
                            currentType = Type.OBJECT;
                        } else {
                            ((List<Object>) currentContainer).add(value);
                            currentType = Type.ARRAY;

                        }
                    }
                    break;
                case HEURISTIC:
                    while (Constants.isWhitespace(current) && i++ < end) {
                        current = jsonString.charAt(i);
                    }

                    if (current != ':' && expectingColon) {
                        stateStack.push(new State(propertyName, currentContainer, Type.OBJECT));
                        throw new JsonParseException(stateStack, "wasn't followed by a colon");
                    }

                    if (current == ':') {
                        if (expectingColon) {
                            expectingColon = false;
                            i++;
                        } else {
                            stateStack.push(new State(propertyName, currentContainer, Type.OBJECT));
                            throw new JsonParseException(stateStack, "was followed by too many colons");
                        }
                    } else if (current == '"') {
                        currentType = Type.STRING;
                        fieldStart = i;
                    } else if (current == '{') {
                        stateStack.push(new State(propertyName, currentContainer, Type.OBJECT));
                        currentType = Type.OBJECT;
                        currentContainer = new HashMap<>();
                        i++;
                    } else if (current == '[') {
                        stateStack.push(new State(propertyName, currentContainer, Type.OBJECT));
                        currentType = Type.ARRAY;
                        currentContainer = new ArrayList<>();
                        i++;
                    } else if (Constants.isLetter(current)) {
                        // Assume parsing a constant ("null", "true", "false", etc)
                        currentType = Type.CONSTANT;
                        fieldStart = i;
                    } else if (Constants.isNumberStart(current)) {
                        // Is a number
                        currentType = Type.NUMBER;
                        fieldStart = i;
                    } else {
                        throw new JsonParseException(stateStack, "unexpected character \"" + current +
                                "\" instead of object value");
                    }
                    break;
                case OBJECT:
                    while (Constants.isWhitespace(current) && i++ < end) {
                        current = jsonString.charAt(i);
                    }

                    if (current == ',') {
                        if (expectingComma) {
                            expectingComma = false;
                            i++;
                        } else {
                            stateStack.push(new State(propertyName, currentContainer, Type.OBJECT));
                            throw new JsonParseException(stateStack, "followed by too many commas");
                        }
                    } else if (current == '"') {
                        if (expectingComma) {
                            stateStack.push(new State(propertyName, currentContainer, Type.OBJECT));
                            throw new JsonParseException(stateStack, "wasn't followed by a comma");
                        }

                        currentType = Type.NAME;
                        fieldStart = i;
                    } else if (current == '}') {
                        if (!stateStack.isEmpty()) {
                            State upper = stateStack.pop();
                            Object upperContainer = upper.container;
                            String parentName = upper.propertyName;
                            currentType = upper.type;

                            if (upperContainer instanceof Map) {
                                ((Map<String, Object>) upperContainer).put(parentName, currentContainer);
                            } else {
                                ((List<Object>) upperContainer).add(currentContainer);
                            }
                            currentContainer = upperContainer;
                            expectingComma = true;
                            i++;
                        } else {
                            return currentContainer;
                        }
                    } else if (!Constants.isWhitespace(current)) {
                        throw new JsonParseException(stateStack, "unexpected character '" + current +
                                "' where a property name is expected. Missing quotes?");
                    }
                    break;
                case ARRAY:
                    while (Constants.isWhitespace(current) && i++ < end) {
                        current = jsonString.charAt(i);
                    }

                    if (current != ',' && current != ']' && current != '}' && expectingComma) {
                        stateStack.push(new State(null, currentContainer, Type.ARRAY));
                        throw new JsonParseException(stateStack, "wasn't preceded by a comma");
                    }

                    if (current == ',') {
                        if (expectingComma) {
                            expectingComma = false;
                            i++;
                        } else {
                            stateStack.push(new State(null, currentContainer, Type.ARRAY));
                            throw new JsonParseException(stateStack, "preceded by too many commas");
                        }
                    } else if (current == '"') {
                        currentType = Type.STRING;
                        fieldStart = i;
                    } else if (current == '{') {
                        stateStack.push(new State(null, currentContainer, Type.ARRAY));
                        currentType = Type.OBJECT;
                        currentContainer = new HashMap<>();
                        i++;
                    } else if (current == '[') {
                        stateStack.push(new State(null, currentContainer, Type.ARRAY));
                        currentType = Type.ARRAY;
                        currentContainer = new ArrayList<>();
                        i++;
                    } else if (current == ']') {
                        if (!stateStack.isEmpty()) {
                            State upper = stateStack.pop();
                            Object upperContainer = upper.container;
                            String parentName = upper.propertyName;
                            currentType = upper.type;

                            if (upperContainer instanceof Map) {
                                ((Map<String, Object>) upperContainer).put(parentName, currentContainer);
                            } else {
                                ((List<Object>) upperContainer).add(currentContainer);
                            }
                            currentContainer = upperContainer;
                            expectingComma = true;
                            i++;
                        } else {
                            return currentContainer;
                        }
                    } else if (Constants.isLetter(current)) {
                        // Assume parsing a   ("null", "true", "false", etc)
                        currentType = Type.CONSTANT;
                        fieldStart = i;
                    } else if (Constants.isNumberStart(current)) {
                        // Is a number
                        currentType = Type.NUMBER;
                        fieldStart = i;
                    } else {
                        stateStack.push(new State(propertyName, currentContainer, Type.ARRAY));
                        throw new JsonParseException(stateStack, "Unexpected character \"" + current + "\" instead of array value");
                    }
                    break;
            }
        }

        throw new JsonParseException("Root element wasn't terminated correctly (Missing ']' or '}'?)");
    }

    private static ExtractedString extractString(String jsonString, int fieldStart) {
        StringBuilder builder = new StringBuilder();
        while (true) {
            int i = indexOfSpecial(jsonString, fieldStart);
            char c = jsonString.charAt(i);
            if (c == '"') {
                builder.append(jsonString.substring(fieldStart + 1, i));
                ExtractedString val = new ExtractedString();
                val.sourceEnd = i;
                val.str = builder.toString();
                return val;
            } else if (c == '\\') {
                builder.append(jsonString.substring(fieldStart + 1, i));

                c = jsonString.charAt(i + 1);
                switch (c) {
                    case '"':
                        builder.append('\"');
                        break;
                    case '\\':
                        builder.append('\\');
                        break;
                    case '/':
                        builder.append('/');
                        break;
                    case 'b':
                        builder.append('\b');
                        break;
                    case 'f':
                        builder.append('\f');
                        break;
                    case 'n':
                        builder.append('\n');
                        break;
                    case 'r':
                        builder.append('\r');
                        break;
                    case 't':
                        builder.append('\t');
                        break;
                    case 'u':
                        builder.append(Character.toChars(
                                Integer.parseInt(jsonString.substring(i + 2, i + 6), 16)));
                        fieldStart = i + 5; // Jump over escape sequence and code point
                        continue;

                }
                fieldStart = i + 1; // Jump over escape sequence
            } else {
                throw new IndexOutOfBoundsException();
            }
        }
    }

    static class State {
        final String propertyName;
        final Object container;
        final Type type;

        State(String propertyName, Object container, Type type) {
            this.propertyName = propertyName;
            this.container = container;
            this.type = type;
        }
    }

    /**
     * Returns the index of either a quotation, or a control character backslash. Skips the first element.
     * !! Do not inline this function, the JVM <3 optimising it, and inlining it slows it down ... somehow.
     * @param str content string to find quote or backslash
     * @param start start index to search
     * @return index of the first quote or backslash found at or after `start`
     */
    private static int indexOfSpecial(String str, int start) {
        while (++start < str.length() && str.charAt(start) != '"' && str.charAt(start) != '\\');
        return start;
    }
    private enum Type {
        ARRAY,
        OBJECT,
        HEURISTIC,
        NAME,
        STRING,
        NUMBER,
        CONSTANT

    }

    private static class ExtractedString {
        int sourceEnd;
        String str;
    }

    public static void main(String[] args) {
        int iterations = Integer.parseInt(args[0]);
        String implementation = args[1];

        String[] toParse = new String[iterations];
        for (int i = 0; i < toParse.length; i++) {
            String uuid = UUID.randomUUID().toString();
            Boolean bool = Math.floor(Math.random() * 2) == 0;
            double rand = Math.random();
            toParse[i] = "{\"a\": " + bool + ", \"uuid\": \"" + uuid + "\", \"rand\": " + rand + "}";
        }

        if ("fasterxml".equals(implementation)) {
            long before = System.currentTimeMillis();
            ObjectMapper mapper = new ObjectMapper();

            for (String parse : toParse) {
                try {
                    mapper.readValue(parse, new TypeReference<Map<String, Object>>(){});
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }

            System.out.println(System.currentTimeMillis() - before);
        } else if ("fuzzlesoft".equals(implementation)) {
            long before = System.currentTimeMillis();

            for (String parse : toParse) {
                JsonParse.map(parse);
            }

            System.out.println(System.currentTimeMillis() - before);
        }
    }
}